src/science/synapse/device.cpp

```diff
@@ -159,6 +159,78 @@ auto Device::uri() const -> const std::string& {
   return uri_;
 }
 
+auto Device::get_logs(
+    const std::string& log_level,
+    std::optional<int64_t> since_ms,
+    std::optional<int64_t> start_time_ns,
+    std::optional<int64_t> end_time_ns,
+    std::optional<std::chrono::milliseconds> timeout) -> std::vector<std::string> {
+
+  synapse::LogQueryRequest request;
+  request.set_min_level(log_level_to_pb(log_level));
+
+  if (since_ms.has_value()) {
+    request.set_since_ms(since_ms.value());
+  } else {
+    if (start_time_ns.has_value()) {
```

Comment:
nice!

while we're here, mind checking that start_time_ns is less than end_time_ns, and if it's not, return an error science::status?